### ISSUE TITLE
[HOLD for RN-Modal to support RN 0.65] [Windows only] Cntrl + click doesn't open the link in a new tab for MenuItem component.

### ISSUE THREAD

**Issue by rushatgabhane** [Expensify/App#3455]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Both click and Cltr + click on MenuItem with a link opens the link in new tab.
## Actual Result:
Only click opens the link in new tab.
Cltr + click doesn't do anything. 

The video below demonstrates that only the component MenuItem (with a link) has this problem. 
Text links work as expected.

https://user-images.githubusercontent.com/29673073/121270451-7235a780-c8ca-11eb-833f-13b733bc490e.mp4

## Action Performed:
1. Open settings menu
2. Go to about
3. Cltr + click on "View the code"

## Workaround:
User can click without pressing Cltr.

## Platform:
Where is this issue occurring?

OS: Windows 
Browser: Chrome

 Web ✔
 iOS
 Android
 Desktop App ❔
 Mobile Web

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/166732
**Upwork job link:** https://www.upwork.com/jobs/~01e82d35df4dd24853

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by michaelhaxhiu** [comment]
hmm, @rushatgabhane this works for me on Mac when I do a cmd + click (web and desktop app). Is it possible this is just impacting Windows users?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by rushatgabhane** [comment]
> hmm, @rushatgabhane this works for me on Mac when I do a cmd + click (web and desktop app). Is it possible this is just impacting Windows users?

Yes, looks like it's an issue for windows and linux only.

**Comment by Julesssss** [comment]
it's certainly minor, but I don't see why we shouldn't fix this for consistency. It would annoy me if this occurred on Mac.

**Comment by michaelhaxhiu** [comment]
Cool, thanks for the bud check @Julesssss. 

I'll post it on Upwork and confirm here when that is done. If you'd like to solve it, you can post a solution proposal @rushatgabhane (in which case, you'd be compensated a bonus for finding this enhancement + solving it)

Then, a member of our eng team will review your proposal and give the 👍 or 👎 on proceeding with a PR.

**Comment by michaelhaxhiu** [comment]
Ok, this job is posted to Upwork here - https://www.upwork.com/jobs/~01e82d35df4dd24853

@rushatgabhane feel free to apply for the job in Upwork if you want to propose a solution, and you'll receive a [$150 bonus in addition to payment for the solution](https://github.com/Expensify/Expensify.cash/blob/main/CONTRIBUTING.md#proposing-a-job-that-expensify-hasnt-posted).

**Comment by rushatgabhane** [comment]
@michaelhaxhiu sure! I'm working on a proposal, will apply soon.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Jag96 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by michaelhaxhiu** [comment]
@Jag96 will be able to assist with reviewing your proposal as the next step.

**Comment by rushatgabhane** [comment]
This issue occurs whenever react-native's Linking is used within a Button, Pressable etc. Doesn't happen for Text.
This is reproducible outside E.cash. That makes me believe this is an issue of React-Native.

**Comment by michaelhaxhiu** [comment]
>This is reproducible outside E.cash, and that makes me think it's a problem with react-native.

Oh wow interesting~ In that case, I feel like we should stick with the React-native standard and wait on them to fix it. It seems minor to me, especially since there's a solid workaround.

@Jag96 lmk if you agree with this angle or want to continue with fixing this. 

Either way @rushatgabhane we appreciate you filing this! We love when contributors discover and file jobs. Keep it up!

**Comment by rushatgabhane** [comment]
> > This is reproducible outside E.cash, and that makes me think it's a problem with react-native.
> 
> Oh wow interesting~ In that case, I feel like we should stick with the React-native standard and wait on them to fix it. It seems minor to me, especially since there's a solid workaround.
> 
> @Jag96 lmk if you agree with this angle or want to continue with fixing this.
> 
> Either way @rushatgabhane we appreciate you filing this! We love when contributors discover and file jobs. Keep it up!

Thanks! Would love to be proved wrong tho.

**Comment by michaelhaxhiu** [comment]
I'll raise it in #expensify-open-source too, it's a perfect forum for input on this.

**Comment by parasharrajat** [comment]
@michaelhaxhiu I would like to fix this.
1. For the web, rendering a Text or view , result in a div to be rendered via `react-native-web`. but Ctrl+click is behaviour for `a`.
So to fix this.
we need to use the `accessibilityRole` with value 'link' which would be converted to `a` on web. Now to make it work properly we have to use the href attribute which would the url which are opening using `Linkiing.openUrl`.
2. Now we have two naviagtion handlers'
  a) one from the Linking.openUrl which is called via onPress handlers.
  b) native browser based due to a with hrefs.
 Both are required but for based on platforms.
 
 Native devices need onpress handlers and they don't understand hrefs.
 Web and M.web and Desktop app can understand a with hrefs.
 
3.  So we will disable the press handlers on Web , Mweb and desktop.
 
4. Now to open the link into new tab we just need to add `hrefAttrs={{target: '_blank'}}`.


*Lets walk the code*

1. We can pass href and hrefAttrs here in options.
https://github.com/Expensify/Expensify.cash/blob/fad0d88c8d23f2b3b91e5ce5c032384210f0ef89/src/pages/settings/AboutPage.js#L42-L46

2. Now we need to ignore the onpress handlers in MenuItems here
https://github.com/Expensify/Expensify.cash/blob/fad0d88c8d23f2b3b91e5ce5c032384210f0ef89/src/components/MenuItem.js#L53

3. we will modify the Pressable here https://github.com/Expensify/Expensify.cash/blob/fad0d88c8d23f2b3b91e5ce5c032384210f0ef89/src/components/MenuItem.js#L52 with something like
```
    <Pressable
        onPress={onPress}
        accessibilityRole='link'
        href={href}
        hrefAttrs={hrefAttrs}
        style={({hovered, pressed}) => ([
            styles.createMenuItem,
            getButtonBackgroundColorStyle(getButtonState(hovered, pressed)),
            wrapperStyle,
        ])}
    >


```

I have tested it and It works like charm and as an advantage, it makes link accessible as you get link suggestions when hovering over them plus readers.

**Comment by parasharrajat** [comment]
> This issue occurs whenever react-native's Linking is used within a Button, Pressable etc. Doesn't happen for Text.
This is reproducible outside E.cash, and that makes me think it's a problem with react-native.

I always remember a mantra that react-native was never created for the web and in the case of the web, it's like a child who needs a guardian `react-native-web`. So if you find an issue on the web, Shout on 'react-native-web' and if it cries and blames `react-native` then go for `react-native`.

**Comment by Jag96** [comment]
Could this be fixed by upgrading react-native-web instead? https://github.com/necolas/react-native-web/issues/2000

**Comment by michaelhaxhiu** [comment]
Slack convo in #expensify-open-source happening [**here**](https://expensify.slack.com/archives/C01GTK53T8Q/p1623442736231300).

**Comment by michaelhaxhiu** [comment]
I like that idea @Jag96. Is there a gh for upgrading to react-native web yet? I'd like to link it here.

**Comment by Jag96** [comment]
Looking at https://expensify.slack.com/archives/C01GTK53T8Q/p1623444561234500?thread_ts=1623442736.231300&cid=C01GTK53T8Q and https://github.com/Expensify/Expensify.cash/pull/3215, it seems that updating to 16 has some changes that will cause problems, so I suggest we hold off on updating this until [this discussion](https://github.com/Expensify/Expensify.cash/pull/3215#issuecomment-859644137) is resolved

**Comment by Jag96** [comment]
From [this comment](https://github.com/Expensify/Expensify.cash/pull/3215#issuecomment-860784429), it appears that we aren't going to be reverting the react-native-web and react package updates, so I think we can resume looking into updating react-native-web in this issue.

cc @parasharrajat @rushatgabhane

**Comment by Julesssss** [comment]
> it appears that we aren't going to be reverting the react-native-web and react package updates, so I think we can resume looking into updating react-native-web in this issue.

@Jag96 yep, good point. Lets continue with your suggestion

**Comment by rushatgabhane** [comment]
@Jag96 sounds good, react-native-web ^0.16.2 has this issue fixed.
Updating will close this issue.

**Comment by Jag96** [comment]
@rushatgabhane perfect! Since you opened this, would you like to fix the issue and make the upgrade PR + tests?

In general, we'd be looking for this issue to be fixed and to not introduce any other regressions. It also looks like there was a CSS hack that was referenced in https://github.com/Expensify/Expensify.cash/pull/3613#issuecomment-862557388, so if we can test to see if that can be removed as well I think that'd be great!

If you don't want to make the fix, let me know and I'll open this up to other contributors!

**Comment by rushatgabhane** [comment]
Yeah, I'd like to upgrade + test for this issue and css hack.

**Comment by Jag96** [comment]
@rushatgabhane perfect! I've assigned you the issue, please submit a proposal on Upwork (https://www.upwork.com/jobs/~01e82d35df4dd24853) so we can send an offer ([readme](https://github.com/Expensify/Expensify.cash/blob/main/CONTRIBUTING.md#payment-for-contributions)).

**Comment by michaelhaxhiu** [comment]
Waiting on @rushatgabhane to apply on Upwork still. Will check again later today.

**Comment by rushatgabhane** [comment]
@michaelhaxhiu I'm trying to fixing the breaking changes that are introduced in new version.
I'll apply after I can do that, I hope that's fine?

**Comment by rushatgabhane** [comment]
I'm having trouble fixing the breaking changes.
[slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1624467008017300)

The gist of it is, many packages like `RN-modal` and `undrescore.debounce()`
use deprecated method `EventEmitter.removeListener` instead of `EvenEmitter.remove()`

**Comment by rushatgabhane** [comment]
It seems like the subscription will be removed, and use of deprected method in third-party libs don't cause a breaking change.
but there will be ~25 `console.error()` for most interactions :(

https://github.com/necolas/react-native-web/blob/13f7b9e2f45af2761bfdccd763c511992c0be030/packages/react-native-web/src/vendor/react-native/emitter/_EventEmitter.js#L169-L183

I believe these are our options
- upgrade to RN-Web 0.17.0 and live with the console errors.
- don't upgrade, fix this issue with a hack.
- use a fork of RN-web with no `console.error()` for this method.

**Comment by parasharrajat** [comment]
Then we will have to upgrade others lib eventually as this feature is deprecated and will be removed without warnings. Were you able to find any thread related to the above issue which suggests why the deprecation decision was taken if a bunch of commonly used libs doesn't support it yet even with the latest RN?.

**Comment by rushatgabhane** [comment]
@parasharrajat apparently RN-Web deprecated the method based on an unstable version of react-native.
https://github.com/facebook/react-native/commit/d39643b9de11c6b44984166ede34a7f44de76fe5

So it makes sense other libs haven't updated to support the depreciation yet.

**Comment by michaelhaxhiu** [comment]
How is this one coming along? Seems we are blocked on finding a method that works, since `EventEmitter.removeListener` is being deprecated?

**Comment by rushatgabhane** [comment]
@michaelhaxhiu Yeah, we first have to wait for ReactNative to release their stable version 0.65.0 which deprecates `EventEmitter.removeListener`.

Then wait other third party libs to upgrade to RN 0.65.0.

This could take a few months easily.

**Comment by parasharrajat** [comment]
Seems like You got it wrong @rushatgabhane. RN-modal does not depend on RN's specific version. You have to upgrade our codebase to RN 0.65.0 from 0.64.

**Comment by rushatgabhane** [comment]
@parasharrajat 
Yeah I could be, but here is an example of RN-Modal using `EventEmitter.removeListener`

https://github.com/react-native-modal/react-native-modal/blob/60ea392df7d26a598628b1f3a06440b7ecccacd9/src/modal.tsx#L259

**Comment by parasharrajat** [comment]
@michaelhaxhiu Based upon n the @rushatgabhane comments above, what do you think about the issue at hand?

We already have another solution https://github.com/Expensify/Expensify.cash/issues/3455#issuecomment-859892539 in case we want it to be fixed now.

**Comment by rdjuric** [comment]
@rushatgabhane Haven't you guys decided in the Slack thread to just update react-native-web and ignore the console warnings? `EventEmitter.removeListener` is just deprecated (not removed), updating seems better than adding a workaround.

**Comment by rushatgabhane** [comment]
@rdjuric Yes, but only if we have to deal with console logs for a week. [Reference comment ](https://expensify.slack.com/archives/C01GTK53T8Q/p1624659830143600?thread_ts=1624658192.141600&cid=C01GTK53T8Q)
And I think we'll have to deal with it for months.

**Comment by michaelhaxhiu** [comment]
@rushatgabhane no problem, this an example of the process working :) Sometimes we need to discuss a job to find the best solution, and this job is a good example!

My understanding is that this should be fixed in the future by the RN-modal version upgrade, but the timeline on that upgrade happening is indefinite. If we want to fix it sooner, @parasharrajat has provided a [proposal](https://github.com/Expensify/Expensify.cash/issues/3455#issuecomment-859892539) that will accomplish it. 

@Jag96 how do you feel about proceeding with @parasharrajat's solution?

**Comment by Jag96** [comment]
A couple of questions on that proposal just to help me understand better:

> Now we need to ignore the onpress handlers in MenuItems here

How do we do that? Do we remove the onPress handler?

> We can pass href and hrefAttrs here in options.

Does that mean that anywhere else we'd want this to work, we'd have to also pass those options to each menu item? So something like this?

```
const menuItems = [
        {
            translationKey: 'initialSettingsPage.aboutPage.appDownloadLinks',
            icon: Link,
            action: () => {
                Navigation.navigate(ROUTES.SETTINGS_APP_DOWNLOAD_LINKS);
            },
            href: some_href,
            hrefAttrs: {target: '_blank'}
        },
...
```

I'm leaning towards not adding any fix for this right now and holding off until RN and the required libraries update everything. This is a pretty low priority and I don't want to introduce hacks if we can avoid it. Going to cc @tgolen to get his thoughts here as well since he was involved in the [Slack convo](https://expensify.slack.com/archives/C01GTK53T8Q/p1624658192141600).

**Comment by parasharrajat** [comment]
> How do we do that? Do we remove the onPress handler?

Just return early

> Does that mean that anywhere else we'd want this to work, we'd have to also pass those options to each menu item? So something like this?.

Yes, I guess we need to pass the Route as href which is passed in `Navigation.navigate`. hrefAttrs are optional and we will are less likely to change them.

 I think we have to upgrade our Codebase to > RN 0.65 as well where a new EventEmitter implementation exists. I am also in favour of a lib upgrade. Maybe we can just ignore those errors using Log box.

**Comment by tgolen** [comment]
> holding off until RN and the required libraries update everything. This is a pretty low priority and I don't want to introduce hacks if we can avoid it

100% agree with this.

**Comment by Jag96** [comment]
Sounds good, going to put a hold on this until RN is at v0.65 and RN-modal supports that version (you can see the currently supported version [here](https://github.com/react-native-modal/react-native-modal/blob/master/package.json#L75)).

**Comment by mallenexpensify** [comment]
Add `Planning` label since it's on hold, to help with GH searching by label

**Comment by MelvinBot** [comment]
@Jag96, @rushatgabhane it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by rushatgabhane** [comment]
Issue is on hold.
Ignore the bot.

**Comment by kidroca** [comment]
Overall I agree with what @parasharrajat is suggesting here: https://github.com/Expensify/App/issues/3455#issuecomment-859892539

---

The semantic and accessibility compliant way is to render an anchor tag
There are 2 types of links 
- internal -> links that take you somewhere else inside the app
- external -> links that take you to an external page

### Internal Links 
We're not handling internal routing as links, but react to button presses (simulating link behavior)
There's nothing wrong with that, but it means some of the default context actions, like opening a LHN conversation in a new tab, aren't possible 
React Navigation provides a guide on how to solve this problem for internal links if we ever reconsider: https://reactnavigation.org/docs/link/
> The Link component lets us navigate to a screen using a path instead of a screen name based on the linking options. It preserves the default behavior of anchor tags in the browser such as Right click -> Open link in new tab", Ctrl+Click/⌘+Click etc.
> If you want to use your own custom touchable, you can use useLinkProps instead

### External Links
It seems our concern for this ticket is regarding external links 

From the `react-native-web` docs we can see that we can make any `View` or `Text` be rendered as a link by providing a `href` prop: https://necolas.github.io/react-native-web/docs/accessibility/#links
We're already doing this with our `TextLink` component 

A `MenuItem` can act more like a link if we provide a `href` and add it to the underlying `<Pressable>`.
This exposes the context menu for links and covers middle click and ctrl clicks 
To incorporate this with less changes and less platform specific logic we can prevent the default behavior of an event - `e.preventDefault()` 
This way we get the best of both worlds - context based actions for links and letting our app decided what happens for the `onPress` handler like calling `openExternalLink`

Since I've already tried this I've captured a draft of the necessary changes here: #6296

With a bit more changes we could also cover internal links

**Comment by Jag96** [comment]
@kidroca that's a great breakdown, thanks for having a look at that. Since this is just affecting the control + click for windows users, I still don't think it's worth making all these changes that can affect how links work on all platforms across the app, so I'm leaning towards leaving this on hold.

If anyone has different opinions, feel free to chime in here.

**Comment by tgolen** [comment]
I'm fine with that stance. Maybe it turns into a bigger issue at some point
and we can revisit this proposal.

**Comment by Jag96** [comment]
Still on hold

**Comment by parasharrajat** [comment]
@Jag96 This should have been resolved now. We upgraded RN-web.

**Comment by rushatgabhane** [comment]
Wohoo! This issue is resolved and can be closed.
But yeah console errors are a thing now.

https://user-images.githubusercontent.com/29673073/148408049-9190e0b7-95f0-4538-a6f1-87c62b2aabe9.mp4

**Comment by mallenexpensify** [comment]
Is any compensation due for this issue?

**Comment by rushatgabhane** [comment]
Hmm, issue reporting bonus? Depends on if it works retroactively.

**Comment by mallenexpensify** [comment]
@rushatgabhane if you reported a bug or feature request that got fixed internally or externally, and you were the first to report it, I think you're eligible for the reporting bonus (even if it's many months old). Sound good @Jag96 ?

**Comment by Jag96** [comment]
Yeah that sounds good to me! @rushatgabhane I'll create an Upwork job and hire you for this one.

**Comment by rushatgabhane** [comment]
Done, thank you!

**Comment by Jag96** [comment]
Paid!
